substrate/frame/revive/src/vm/evm/instructions/host.rs

```diff
@@ -143,18 +98,18 @@ pub fn extcodecopy<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 ///
 /// Gets the hash of one of the 256 most recent complete blocks.
 pub fn blockhash<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	gas_legacy!(context.interpreter, gas::BLOCKHASH);
+	gas!(context.interpreter, RuntimeCosts::BlockHash);
 	popn_top!([], number, context.interpreter);
+	let requested_number = <sp_core::U256 as U256Converter>::from_revm_u256(&number);
 
-	let requested_number = *number;
-	let block_number = context.host.block_number();
+	let block_number = context.interpreter.extend.block_number();
 
 	let Some(diff) = block_number.checked_sub(requested_number) else {
 		*number = U256::ZERO;
 		return;
 	};
 
-	let diff = as_u64_saturated!(diff);
+	let diff = if diff > sp_core::U256::from(u64::MAX) { u64::MAX } else { diff.low_u64() };
```

Comment:
Logic needs to be moved into the `Ext`. Everything that can be in ext should live there.